Sources/ConsoleKit/Utilities/LoggerFragment.swift

```diff
@@ -0,0 +1,307 @@
+        self.literal = literal
+        self.fragment = fragment
+    }
+    
+    public func write(_ record: inout LogRecord, to output: inout FragmentOutput) {
+        if output.needsSeparator {
+            if self.fragment.hasContent(record: &record) {
+                output.needsSeparator = false
+                output += self.literal
+            }
+        }
+        
+        self.fragment.write(&record, to: &output)
+    }
+}
+
+/// Writes the logged message to the output, and requests a separator for the next fragment.
+public struct MessageFragment: LoggerFragment {
+    public init() { }
+    
+    public func write(_ record: inout LogRecord, to output: inout FragmentOutput) {
+        output += record.message.description.consoleText()
+        output.needsSeparator = true
+    }
+}
+
+/// Writes the combined metadata to the output, and requests a separator for the next fragment only if the metadata was not empty.
+///
+/// This fragment is considered to not have content if the metadata is empty.
+public struct MetadataFragment: LoggerFragment {
+    public init() { }
+    
+    public func hasContent(record: inout LogRecord) -> Bool {
+        !record.allMetadata().isEmpty
+    }
+    
+    public func write(_ record: inout LogRecord, to output: inout FragmentOutput) {
+        let allMetadata = record.allMetadata()
+        
+        guard !allMetadata.isEmpty else { return }
+        
+        output += allMetadata.sortedDescriptionWithoutQuotes.consoleText()
+        output.needsSeparator = true
+    }
+}
+
+/// Writes the file location of the logged message, including the line. This fragment requests a separator for the next fragment.
+public struct SourceLocationFragment: LoggerFragment {
+    public init() { }
+    
+    public func write(_ record: inout LogRecord, to output: inout FragmentOutput) {
+        let file = record.file + ":" + record.line.description
+        output += "(" + file.consoleText() + ")"
+        output.needsSeparator = true
+    }
+}
+
+public struct TimestampFragment: LoggerFragment {
+    public init() { }
+    
+    public func write(_ record: inout LogRecord, to output: inout FragmentOutput) {
+        output += self.timestamp().consoleText()
+        output.needsSeparator = true
+    }
+    
+    private func timestamp() -> String {
+        var buffer = [Int8](repeating: 0, count: 255)
+#if os(Windows)
+        var timestamp = __time64_t()
+        _ = _time64(&timestamp)
+        
+        var localTime = tm()
+        _ = _localtime64_s(&localTime, &timestamp)
+        
+        _ = strftime(&buffer, buffer.count, "%Y-%m-%dT%H:%M:%S%z", &localTime)
+#else
+        var timestamp = time(nil)
+        let localTime = localtime(&timestamp)
+        strftime(&buffer, buffer.count, "%Y-%m-%dT%H:%M:%S%z", localTime)
+#endif
+        return buffer.withUnsafeBufferPointer {
+            $0.withMemoryRebound(to: CChar.self) {
+                String(cString: $0.baseAddress!)
+            }
+        }
```

Comment:
I forgot to suggest before to replace this with this much more efficient (and, on macOS and Linux, thread-safe) version:
```suggestion
        withUnsafeTemporaryAllocation(of: CChar.self, capacity: 255) {
            #if os(Windows)
                var timestamp = __time64_t()
                var localTime = tm()
                _ = _time64(&timestamp)
                _ = _localtime64_s(&localTime, &timestamp)
            #else
                var timestamp = time(nil)
                var localTime = tm()
                localtime_r(&timestamp, &localTime)
            #endif
            guard strftime($0.baseAddress!, $0.count, "%Y-%m-%dT%H:%M:%S%z", &localTime) > 0 else {
                return "<unknown>"
            }
            return String(cString: $0.baseAddress!)
        }
```